floodmodeller_api/units/_base.py

```diff
@@ -99,3 +107,120 @@ def __eq__(self, other, return_diff=False):
+
+    # rules & varrules
+
+    def _read_rules(self, block):
+        rule_params = split_10_char(block[self._last_gate_row + 1])
+        self.nrules = int(rule_params[0])
+        self.rule_sample_time = _to_float(rule_params[1])
+        self.timeunit = _to_str(rule_params[2], "SECONDS", check_float=False)
+        self.extendmethod = _to_str(rule_params[3], "EXTEND")
+        self.rules = self._get_logical_rules(
+            self.nrules, block, self._last_gate_row + 2
+        )
+        # Get time rule data set
+        nrows = int(split_10_char(block[self._last_rule_row + 1])[0])
+        data_list = []
+        for row in block[self._last_rule_row + 2 : self._last_rule_row + 2 + nrows]:
+            row_split = split_10_char(f"{row:<20}")
+            x = _to_float(row_split[0])  # time
+            y = row[10:].strip()  # operating rules
+            data_list.append([x, y])
+        self._last_time_row = self._last_rule_row + nrows + 1
+        rule_data = pd.DataFrame(data_list, columns=["Time", "Operating Rules"])
+        rule_data = rule_data.set_index("Time")
+        rule_data = rule_data["Operating Rules"]
+        self.time_rule_data = rule_data
+        # VARRULES (not always necessary)
+        self.has_varrules = False
+        if self._last_time_row + 1 < len(block):
+            if block[self._last_time_row + 1].strip() == "VARRULES":
+                self.has_varrules = True
+                varrule_params = split_10_char(block[self._last_time_row + 2])
+                self.nvarrules = int(varrule_params[0])
+                self.varrule_sample_time = _to_float(rule_params[1])
+                self.varrules = self._get_logical_rules(
+                    self.nvarrules, block, self._last_time_row + 3
+                )
+                # Get time rule data set
+                var_nrows = int(split_10_char(block[self._last_rule_row + 1])[0])
+                data_list = []
+                for row in block[
+                    self._last_rule_row + 2 : self._last_rule_row + 2 + var_nrows
+                ]:
+                    row_split = split_10_char(f"{row:<20}")
+                    x = _to_float(row_split[0])  # time
+                    y = row[10:].strip()  # operating rules
+                    data_list.append([x, y])
+
+                varrule_data = pd.DataFrame(
+                    data_list, columns=["Time", "Operating Rules"]
+                )
+                varrule_data = varrule_data.set_index("Time")
+                varrule_data = varrule_data["Operating Rules"]
+                self.time_varrule_data = varrule_data
+
+    def _write_rules(self, block):
+        # ADD RULES
+        block.append("RULES")
+        self.nrules = len(self.rules)
+        block.append(
+            f"{join_n_char_ljust(10, self.nrules)}{join_10_char(self.rule_sample_time)}{join_n_char_ljust(10, self.timeunit, self.extendmethod)}"
+        )
+        for rule in self.rules:
```

Comment:
I think we should probably catch an exception here for if a rule has been added that doesn't end with "END"